unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py

```diff
@@ -0,0 +1,363 @@
+                "type": "object",
+                "properties": {
+                    "name": {"type": ["null", "string"]},
+                    "age": {"type": ["null", "integer"]},
+                },
+            },
+        ),
+        (
+            # Test case: Fields with missing type default to "string"
+            iter(
+                [
+                    {
+                        "schema": [
+                            {"key": "name"},
+                            {"key": "email", "type": "string"},
+                        ]
+                    }
+                ]
+            ),
+            {
+                "$schema": "http://json-schema.org/draft-07/schema#",
+                "type": "object",
+                "properties": {
+                    "name": {"type": ["null", "string"]},
+                    "email": {"type": ["null", "string"]},
+                },
+            },
+        ),
+        (
+            # Test case: Fields with nested types
+            iter(
+                [
+                    {
+                        "schema": [
+                            {"key": "address", "type": ["string", "integer"]},
+                        ]
+                    }
+                ]
+            ),
+            {
+                "$schema": "http://json-schema.org/draft-07/schema#",
+                "type": "object",
+                "properties": {
+                    "address": {
+                        "oneOf": [{"type": ["null", "string"]}, {"type": ["null", "integer"]}]
+                    },
+                },
+            },
+        ),
+        (
+            # Test case: Empty record set
+            iter([]),
+            {
+                "$schema": "http://json-schema.org/draft-07/schema#",
+                "type": "object",
+                "properties": {},
+            },
+        ),
+    ],
+)
+def test_dynamic_schema_loader(dynamic_schema_loader, retriever_data, expected_schema):
+    dynamic_schema_loader.retriever.read_records = MagicMock(return_value=retriever_data)
+
+    schema = dynamic_schema_loader.get_json_schema()
+
+    # Validate the generated schema
+    assert schema == expected_schema
+
+
+def test_dynamic_schema_loader_invalid_key(dynamic_schema_loader):
+    # Test case: Invalid key type
+    dynamic_schema_loader.retriever.read_records.return_value = iter(
+        [{"schema": [{"field1": {"key": 123, "type": "string"}}]}]
+    )
+
+    with pytest.raises(ValueError, match="Expected key to be a string"):
+        dynamic_schema_loader.get_json_schema()
+
+
+def test_dynamic_schema_loader_invalid_type(dynamic_schema_loader):
+    # Test case: Invalid type
+    dynamic_schema_loader.retriever.read_records.return_value = iter(
+        [{"schema": [{"field1": {"key": "name", "type": "invalid_type"}}]}]
+    )
+
+    with pytest.raises(ValueError, match="Expected key to be a string. Got None"):
+        dynamic_schema_loader.get_json_schema()
+
+
+#
```

Comment:
It feels weird to have this comment here. We could have a `test_integration_dynamic_schema_loader.py` but I think I have a slight preference on having all those tests in this file

Reply:
I feel the same for the imports, which should be moved above, but I think it was just a lousy copy-paste.

Reply:
These are artifacts from testing. My bad. They have been removed.